src/main/scala/no/gdl/bookapi/service/search/IndexBuilderService.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Part of GDL book_api.
+ * Copyright (C) 2017 Global Digital Library
+ *
+ * See LICENSE
+ */
+
+package no.gdl.bookapi.service.search
+
+import com.typesafe.scalalogging.LazyLogging
+import io.digitallibrary.language.model.LanguageTag
+import no.gdl.bookapi.BookApiProperties
+import no.gdl.bookapi.model.domain.{ReindexResult, Translation}
+import no.gdl.bookapi.repository.TranslationRepository
+
+import scala.util.{Failure, Success, Try}
+
+trait IndexBuilderService {
+  this: TranslationRepository with IndexService =>
+  val indexBuilderService: IndexBuilderService
+
+  class IndexBuilderService extends LazyLogging {
+
+    def indexDocument(imported: Translation): Try[Translation] = {
+      for {
+        _ <- indexService.aliasTarget(imported.language).map {
+          case Some(index) => Success(index)
+          case None => indexService.createSearchIndex(imported.language).map(newIndex => indexService.updateAliasTarget(None, newIndex, imported.language))
+        }
+        imported <- indexService.indexDocument(imported)
+      } yield imported
+    }
+
+    def indexDocuments: Try[ReindexResult] = {
+      synchronized {
+        var numIndexed = 0;
+        val start = System.currentTimeMillis()
+        getLanguages.map(languages => {
+          languages.foreach(language => {
+            indexService.createSearchIndex(language).flatMap(indexName => {
+              val operations = for {
+                numIndexed <- sendToElastic(indexName, language)
+                aliasTarget <- indexService.aliasTarget(language)
+                _ <- indexService.updateAliasTarget(aliasTarget, indexName, language)
+                _ <- indexService.deleteSearchIndex(aliasTarget)
+              } yield numIndexed
+
+              operations match {
+                case Failure(f) => {
+                  indexService.deleteSearchIndex(Some(indexName))
+                  Failure(f)
+                }
+                case Success(totalIndexed) => {
+                  Success(numIndexed += totalIndexed)
+                }
+              }
+            })
+          })
+        })
+        Success(ReindexResult(numIndexed, System.currentTimeMillis() - start))
+      }
+    }
+
+    def sendToElastic(indexName: String, languageTag: LanguageTag): Try[Int] = {
+      var numIndexed = 0
```

Comment:
This isn't the most idiomatic way to do this, as it mutates `numIndexed`. We could try to do this in a pure manner.

Reply:
Need to do operation x number of timed, and at the same time add up total indexed. This seemed like a ok way to do this.